app/gui/src/project-view/components/visualizations/ScatterplotVisualization.vue

```diff
@@ -697,90 +701,114 @@ function getTooltipMessage(point: Point) {
       point.series && point.series in axis ?
         axis[point.series as keyof AxesConfiguration].label
       : ''
-    return `${formatXPoint(point.x)}, ${point.y}, ${label}- Double click to inspect point`
+    return `${formatXPoint(point.x)}, ${point.y}, ${label} - Double click to inspect point`
   }
-  return `${formatXPoint(point.x)}, ${point.y}- Double click to inspect point`
+  return `${formatXPoint(point.x)}, ${point.y} - Double click to inspect point`
 }
 
 // === Update contents ===
 
 watchPostEffect(() => {
   const xScale_ = data.value.isTimeSeries ? xScaleTime.value : xScale.value
   const yScale_ = yScale.value
-  const plotData = getPlotData(data.value) as Point[]
+
+  const allPlotData = getPlotData(data.value) as Point[]
+  const circleData = allPlotData.filter((p) => (p.shape || 'circle') === 'circle')
+  const symbolData = allPlotData.filter((p) => (p.shape || 'circle') !== 'circle')
```

Comment:
We have a `partition` function to split an array into two buckets based on bool predicate. You can import it from `util/data/array`. That should reduce this to a single pass over the data set.

```suggestion
  const [symbolData, circleData] = partition(allPlotData, (p) => p.shape !== 'circle')
```